### ISSUE TITLE
Problems with ReactDOM.findDOMNode()

### ISSUE THREAD

**Issue by PolGuixe** [airbnb/enzyme#479]
I am getting this error: 
```
Invariant Violation: Element appears to be neither ReactComponent nor DOMNode (keys: value)
```
When I am trying to simulate a submit event which its assoaciated function uses `ReactDOM.findDOMNode(ref).value` to recover the input value. 

----------
More info. 
```
//Inside the form
<FormControl type="text" ref='titleRef' placeholder="Enter your product name"/> //I am using React-Boostrap

//Inside the onSubmit function
const {titleRef, contentRef} = this.refs;
save(ReactDOM.findDOMNode(titleRef).value, ReactDOM.findDOMNode(contentRef).value);

Any idea why it isn't working?

**Comment by PolGuixe** [comment]
It is also being discussed here: https://github.com/react-bootstrap/react-bootstrap/issues/1996#issuecomment-229698258

**Comment by blainekasten** [comment]
Are you using `shallow` or `mount`?

**Comment by PolGuixe** [comment]
I have tried with both and I am getting the same error.

**Comment by blainekasten** [comment]
Try removing the `ReactDOM.findDOMNode` code and just acessing directly.

```js
this.refs.contentRef.value;
this.refs.titleRef.value;

**Comment by PolGuixe** [comment]
I know the test approach works, because I have tried it with a different form in which I am not using React-Boostrap components, hence I can do `ref.value` to retrieve the value from the input, and then the test passes.

**Comment by blainekasten** [comment]
Word. I believe you. It's hard to pinpoint where it is. Can you share a gist of the underlying component and the test code?

**Comment by PolGuixe** [comment]
@blainekasten here it is https://gist.github.com/PolGuixe/5e07c7c6cf9fc7c19bc7444bf0d66373

**Comment by blainekasten** [comment]
I'm going to strongly guess this is your problem.

https://gist.github.com/PolGuixe/5e07c7c6cf9fc7c19bc7444bf0d66373#file-product_edit-tests-js-L26-L33

If you look at `findDOMNode` code [here](https://github.com/facebook/react/blob/master/src/renderers/dom/client/findDOMNode.js) it is trying to find the component in some sort of internal map. I don't think you're going to be able to mock out that properly and stable-enough for this approach to be trustworthy. 

I would say if you are wanting these refs to have certain values, set them with native code.

Something like this might work:

```js
const wrapper = mount(<ProductEdit />);
wrapper.find('FormControl').at(0).setProps({value: name});
// set other one
// simulate
```

**Comment by PolGuixe** [comment]
Thanks! I am trying to implement this approach. However `setProps` is triggering an error: `Error: ReactWrapper::setProps() can only be called on the root`.

I guess that setProps can only be used with the `wrapper` itself.

**Comment by blainekasten** [comment]
ah yeah. maybe that's not the way.. Somehow you need to set the value of the node itself. I can't remember the exact way. But I think you can access the internal node via `wrapper.node`.

So maybe something like `find('input').node.value = foo`.

I'm going to close this issue as it doesn't seem to be an Enzyme problem. But feel free to keep talking if you need assistance! Glad we figured it out-ish.

**Comment by PolGuixe** [comment]
I am struggling a bit. 
Doing `find('FormControl').at(0).node.value = foo;` Doesn't work. 
Doing `find('input').at(0).node.value = foo;` Works. 
However, `find('input').at(1).node.value = foo; Doesn't work. //find('input').at(1) is undefined.

Can I find the node by its ref?

**Comment by blainekasten** [comment]
So, doing `.node` is returning the underlying html node. `FormControl` does not have a `.node`. So that wuldn't work. But looking into react-bootstrap, looks like `FormControl` renders an `input`.

So that's why doing the `input.node.value` does work.

But yes, there is a [ref API](https://github.com/airbnb/enzyme/blob/master/docs/api/ReactWrapper/ref.md) you could try out.

**Comment by rimunroe** [comment]
> I think you can access the internal node via wrapper.node.

@blainekasten It looks like `node` is a property of all wrappers, but is it meant to be only used internally? I'm in a similar boat to the submitter with needing to use `findDOMNode`, but I can't find any documentation for this.

 Sorry to pop in to such an old issue.

**Comment by ljharb** [comment]
@rimunroe use `.getNode()` now, rather than `.node` (which we hope to remove eventually)

**Comment by rimunroe** [comment]
Thanks for the prompt response! It appears that  `getNode` doesn't appear in the API docs. Is that intentional?

**Comment by ljharb** [comment]
Hmm, no - it was added in #490 which is released in v2.6.0. We may have missed adding the docs for that.

**Comment by rimunroe** [comment]
> Hmm, no - it was added in #490 which is released in v2.6.0. We may have missed adding the docs for that.

Ah! Thanks!
